File: examples/delta-polling.js
const API_ORIGIN = process.env.X402NANO_API_ORIGIN || "https://x402nano.onrender.com";
const SLUG = process.env.X402NANO_MARKET_SLUG || "will-gideon-saar-be-the-next-prime-minister-of-israel";
const POLL_INTERVAL_MS = Number(process.env.X402NANO_POLL_INTERVAL_MS || 5 * 60 * 1000);

async function getJson(path, options = {}) {
  const response = await fetch(new URL(path, API_ORIGIN), options);
  const body = await response.json().catch(() => ({}));
  return { response, body };
}

async function createXPayment(_challenge) {
  // Plug in your x402 wallet/client here. Return the encoded X-PAYMENT value.
  // Keep private keys outside this file, for example in a wallet service or env-backed signer.
  throw new Error("createXPayment() must be connected to an x402 client before paid polling.");
}

function shouldPayForDelta(freshness) {
  if (!freshness?.generatedOnRequest) return false;
  if ((freshness.estimatedLatencyMs?.p95 ?? Infinity) > 5000) return false;
  return freshness.expectedMaxAgeSeconds <= 120;
}

async function pollDeltaOnce(since) {
  const path = `/api/markets/delta?slug=${encodeURIComponent(SLUG)}&since=${encodeURIComponent(since)}`;
  const unpaid = await getJson(path);

  if (unpaid.response.status !== 402) {
    throw new Error(`Expected 402 before payment, got ${unpaid.response.status}`);
  }

  const freshness = unpaid.body.x402?.resource?.freshness;
  if (!shouldPayForDelta(freshness)) {
    return { paid: false, since, reason: "freshness metadata did not meet local policy" };
  }

  const payment = await createXPayment(unpaid.body.x402);
  const paid = await getJson(path, { headers: { "X-PAYMENT": payment } });

  if (!paid.response.ok) {
    throw new Error(paid.body.reason || paid.body.error || `Paid request failed with ${paid.response.status}`);
  }

  const nextSince = paid.body.data?.window?.until || new Date().toISOString();
  return {
    paid: true,
    since: nextSince,
    receipt: paid.body.receipt?.id,
    repeatCheckPriority: paid.body.data?.significance?.repeatCheckPriority,
    summary: paid.body.data?.significance?.summary
  };
}

async function main() {
  let since = process.env.X402NANO_SINCE || new Date(Date.now() - 15 * 60 * 1000).toISOString();

  while (true) {
    try {
      const result = await pollDeltaOnce(since);
      console.log(JSON.stringify(result, null, 2));
      since = result.since;
    } catch (error) {
      console.error(`poll failed: ${error.message}`);
    }

    await new Promise(resolve => setTimeout(resolve, POLL_INTERVAL_MS));
  }
}

main().catch(error => {
  console.error(error);
  process.exitCode = 1;
});

File: examples/delta-polling.py
import json
import os
import time
from datetime import datetime, timedelta, timezone
from urllib.parse import urlencode
from urllib.request import Request, urlopen
from urllib.error import HTTPError

API_ORIGIN = os.getenv("X402NANO_API_ORIGIN", "https://x402nano.onrender.com").rstrip("/")
SLUG = os.getenv("X402NANO_MARKET_SLUG", "will-gideon-saar-be-the-next-prime-minister-of-israel")
POLL_INTERVAL_SECONDS = int(os.getenv("X402NANO_POLL_INTERVAL_SECONDS", "300"))


def get_json(path, headers=None):
    request = Request(f"{API_ORIGIN}{path}", headers=headers or {})
    try:
        with urlopen(request, timeout=15) as response:
            return response.status, json.loads(response.read() or b"{}")
    except HTTPError as error:
        return error.code, json.loads(error.read() or b"{}")


def create_x_payment(_challenge):
    """Plug in your x402 wallet/client here and return the encoded X-PAYMENT value."""
    raise RuntimeError("create_x_payment() must be connected to an x402 client before paid polling.")


def should_pay_for_delta(freshness):
    if not freshness or not freshness.get("generatedOnRequest"):
        return False
    if freshness.get("estimatedLatencyMs", {}).get("p95", 999999) > 5000:
        return False
    return freshness.get("expectedMaxAgeSeconds", 999999) <= 120


def poll_delta_once(since):
    query = urlencode({"slug": SLUG, "since": since})
    path = f"/api/markets/delta?{query}"
    status, unpaid_body = get_json(path)

    if status != 402:
        raise RuntimeError(f"Expected 402 before payment, got {status}")

    freshness = unpaid_body.get("x402", {}).get("resource", {}).get("freshness")
    if not should_pay_for_delta(freshness):
        return {"paid": False, "since": since, "reason": "freshness metadata did not meet local policy"}

    payment = create_x_payment(unpaid_body.get("x402"))
    paid_status, paid_body = get_json(path, headers={"X-PAYMENT": payment})

    if paid_status >= 400:
        raise RuntimeError(paid_body.get("reason") or paid_body.get("error") or f"Paid request failed with {paid_status}")

    next_since = paid_body.get("data", {}).get("window", {}).get("until") or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    return {
        "paid": True,
        "since": next_since,
        "receipt": paid_body.get("receipt", {}).get("id"),
        "repeatCheckPriority": paid_body.get("data", {}).get("significance", {}).get("repeatCheckPriority"),
        "summary": paid_body.get("data", {}).get("significance", {}).get("summary"),
    }


def main():
    since = os.getenv("X402NANO_SINCE") or (datetime.now(timezone.utc) - timedelta(minutes=15)).isoformat().replace("+00:00", "Z")

    while True:
        try:
            result = poll_delta_once(since)
            print(json.dumps(result, indent=2))
            since = result["since"]
        except Exception as error:
            print(f"poll failed: {error}")

        time.sleep(POLL_INTERVAL_SECONDS)


if __name__ == "__main__":
    main()
